sensorlib/src/sensor/SemanticLidarSensor.py

```diff
@@ -260,15 +261,25 @@ def sample_hitpoints(self, hitpoints, sample_size):
         a specific object is less than the sample size, the whole
         hitpoint population will be used (i.e., the object's list of
         hitpoints will remain unchanged).
+        NOTE: sizes and ratio in inputs affect individual object_idx in CARLA 0.9.10
+        which is same as object_tag, but not actor_id. For example, if there are 2 vehicle.carlamotors.carlacola,
+        the size and ratios will apply to total hitpoints of vehicle.carlamotors.carlacola, but not individually.
 
         :param hitpoints: lidar points associated with each object
-        :param sample_size: maximum size of the object's new hitpoint list
+        :param min_sample_size: minimum size to sample from hitpoint list for each objects
+        :param max_sample_size: maximum size to sample from hitpoint list for each objects
+        :param downsample_ratio: int number determining the factor to which the size of the sample is reduced
         :return: downsampled hitpoints
         """
+
+        min_sample_size_ratio_wise = (int)(max(len(points) / downsample_ratio, min_sample_size))
+        sample_size_capped_with_boundary = min(len(points), min_sample_size_ratio_wise)
+        true_sample_size = min(max_sample_size, sample_size_capped_with_boundary)
+
         return {
             # The choice() function will raise an error if we try to
             # sample more than the population
-            id_: self.__rng.choice(points, min(len(points), sample_size), replace=False)
+            id_: self.__rng.choice(points, true_sample_size, replace=False)
             for id_, points in hitpoints.items()
         }
```

Comment:
This implementation seems overly-complex. My understanding is that we have an original population (`population`) that we want to down-sample (`downsampled_poulation`) by some factor (`downsample_factor`). We also want to ensure the resulting population size is neither too big (`max_population_size`) nor too small (`min_population_size`). If all this holds true, it seems like the implementation could be simplified a bit:
```python
def downsample_hitpoints(self, hitpoints, min_size, max_size, factor):
    def clamp(value):
        return min(min_size, max(value, max_size))

    return {
        id_: self.__rng.choice(
            points, clamp(math.ceil(len(points) / factor)), replace=False
        )
        for id_, points in hitpoints
    }

```

Reply:
Yes, that's correct. Good catch.

Reply:
had to add another wrapper `min()` to catch the smaller sizes than small_size